---
title: "SNV analysis"
author: "Sonali Arora"
date: "Nov 16, 2022"
output:
  html_document:
    toc: true
    them

---

# fishHook analysis 

```{r}
rm(list=ls())
library(fishHook)    
library(rtracklayer)
library(org.Hs.eg.db)

expand_df = read.csv("3UTR_Mutations_Hyperremoved_2+_Uncombined.csv", header=T, stringsAsFactors=FALSE)
mutations <- makeGRangesFromDataFrame(expand_df, keep.extra.columns = TRUE)

v22 = import("~/hsiehlab/bioinformatic_references/GATK_hg38_reference/gencode.v22.annotation.gtf")
genes = v22[which(v22$type=="gene"), ]

# calculate the G and C nucleotide frequency.
library(BSgenome.Hsapiens.UCSC.hg38)
genome <- BSgenome.Hsapiens.UCSC.hg38
keep = head(seqlengths(genome), 25) 
gr0 <- GRanges(Rle(c(names(keep))),  
               IRanges(start =1, end =keep))
win = unlist(tile(gr0, width=10001)) 
seq <- getSeq(BSgenome.Hsapiens.UCSC.hg38, win) 
alf <- alphabetFrequency(seq, as.prob=TRUE)[, c("G", "C")] 
head(alf)
colSums(alf)
mcols(win) = alf # add to GRanges object

chromhmm = import('reviewer_questions_March2023/try_hotspot_analysis/BSS01459_18_CALLS_segments.bed.gz') 
hetchromdata = chromhmm %Q% (name %in% c('ZNF/Rpts', 'Het', 'Quies')) 

gc =  fishHook:::Cov(data = win, field = c('C', 'G')) 
hetchrom = Cov(hetchromdata, name = 'Heterochromatin') 

fish3 = fishHook:::Fish(hypotheses = genes,
                        events = mutations,
                        covariates = c( gc, hetchrom),
                        eligible= interval,
                        idcol = 'sampleName',  
                        na.rm = T)

fish3$score()
adv_result <- fish3$res %Q% order(p)
sig_genes2 <- adv_result[which(adv_result$fdr < 0.05),]$gene_name
adv_df <- as.data.frame(adv_result)
write.table(adv_df, file.path("fishHook","fishhook_result_with_gc_content_hetchromm_model_only_utr3_region.txt"), 
            sep="\t", quote=FALSE, row.names=FALSE, col.names=TRUE)
pdf(file.path(resdir, "plot_fishhook_with_gc_content__hetchromm_utr3_region.pdf"), width=10)
fish3$qqp(plotly =FALSE)
dev.off()
```

# prepare input files for pyclone. 

We have a master file for 3' utr , 5' utr and cds mutations, in the below chucnk we create 1 file 
for each patient containing the respective 3' ut, 5' utr and cds mutations from that patient
( see 05 code series for pyclone analysis)

```{r}

rm(list=ls())
library(GenomicRanges)
df = read.csv("3UTR_Mutations_Hyperremoved_2+_Uncombined.csv")
sp = split(df, df$sampleName)

all_nms =c(grep("DTB", names(sp), value=T), 
           "lucap145", "lucap78" , "lucap81" , "lucap92" , 
           "00-010",   "00-029"  , "00-090" ,  "00-140" ,  "00-169" ,  "01_087" , 
           "01-002"  , "01-095" ,  "01-120" ,  "02_083" ,  "02-065" ,  "03_027" , 
           "03_081" ,  "03_082" ,  "03_139" ,  "03-130" ,  "03-163" ,  "03-192" , 
           "04-050" ,  "04-101"  , "04-108" ,  "04-112" ,  "04-149" ,  "05_011" , 
           "05_116" ,  "05_123" ,  "05_214" ,  "05_221" ,  "05-092" ,  "05-148" , 
           "05-187" ,  "05-217" ,  "06_047" ,  "06_081" ,  "06-127" ,  "06-134" , 
           "07_062" ,  "07-042" , "07-050"  , "08-006"  , "08-020"  , "08-037" , 
           "08-093" ,  "09-006" ,  "09-012" , "10-013"  , "10-039"  , "10-056"  ,
           "11_028" ,  "12-005" ,  "12-011" ,  "12-021" ,  "13-012" ,  "13-042" , 
           "13-084" ,  "13-099" ,  "13-101" ,  "13-104" ,  "13-117" ,  "13-122" , 
           "14-031" ,  "14-039" ,  "14-043" ,  "14-053" ,  "14-077" ,  "14-091" ,
           "14-096" ,  "14-105" ,  "15-003" ,  "15-010" ,  "15-023" ,  "15-069" , 
           "15-090" ,  "15-096" ,  "16-052" ,  "16-071"  , "98-372"  , "99-064"  )


cds = read.delim("Master_cds_Mutations_Hyperremoved_2+.txt", header=T, stringsAsFactors=FALSE)
utr5 = read.delim("Master_5UTR_Mutations_Hyperremoved_2+.txt", header=T, stringsAsFactors=FALSE)
cds_nms = unique( trimws(unlist(strsplit(cds$sampleName, ",|.txt"))))
utr5_nms  = unique( trimws(unlist(strsplit(utr5$sampleName, ",|.txt"))))
utr5_by_sample = lapply(utr5_nms, function(x) utr5[grep(x, utr5$sampleName), ])
cds_by_sample = lapply(cds_nms, function(x) cds[grep(x, cds$sampleName), ])
names(cds_by_sample) = cds_nms
names(utr5_by_sample) = utr5_nms

resdir_pyclone = "input_for_pyclone"
sequenza_all_samples= read.delim("sequenza_all_samples.txt", 
                                 header=T, stringsAsFactors = FALSE)

want_cols  = c("Chr",    "Start",  "End", "Ref", "Alt", "Func.refGene", 
               "Gene.refGene", "Total_reads_in_normal", "Ref_reads_in_normal", "Alt_reads_in_normal", "Normal_VAF",
               "Total_reads_in_Tumor", 
               "Ref_reads_in_Tumor" ,"Alt_reads_in_Tumor", "Tumor_VAF", "id")


a1 = lapply(all_nms, function(x){
  utr3_df = sp[[match(x, names(sp))]]
  utr5_df = utr5_by_sample[[match(x, names(utr5_by_sample) )]]
  
  cds_idx = match(x, names(cds_by_sample))
  if(length(cds_idx)==0){
    x = gsub("_", "-", x)
    cds_idx = match(x, names(cds_by_sample))
  }                
  cds_df = cds_by_sample[[cds_idx]]
  
  utr3_df = utr3_df[, match(want_cols, colnames(utr3_df))]
  utr5_df = utr5_df[, match(want_cols, colnames(utr5_df))]
  cds_df = cds_df[, match(want_cols, colnames(cds_df))]
  
  rbind(utr3_df, utr5_df, cds_df)
})

mut_data = list.files(path = resdir, pattern = "_mutations_by_patient.txt", full.names=T)
seq_data = list.files(path ="sequenza", 
             pattern = "segments.txt", recursive=T, full.names=T)

a2 = lapply(1:length(all_nms), function(yidx){
  
  x = all_nms[yidx]
  idx = grep(x, seq_data)
  if(length(idx)==0){
    x = gsub("-", "_", x)
  }                
  seq_fname = seq_data[grep(x, seq_data)]
  
  mut_df = a1[[yidx]]
  seq_df = read.delim(seq_fname, header=T, stringsAsFactors=FALSE)
  seq_df$normal_cn = 2
  
  # fix the chrX and chrY  : 
  xidx = which(seq_df$chromosome %in% c("chrX", "chrY"))
  if(length(xidx)>=1){
    tempdf = seq_df[xidx, ]
    seq_df = seq_df[-xidx, ]
    tempdf$A  = tempdf$CNt
    tempdf$B =0
    tempdf$normal_cn = 1
    seq_df = rbind(seq_df, tempdf)
  }
  
  mut_gr = makeGRangesFromDataFrame(mut_df)
  seq_gr = makeGRangesFromDataFrame(seq_df, start.field="start.pos", end.field="end.pos")
  hits = findOverlaps(mut_gr, seq_gr)
  want_mut = mut_df[queryHits(hits), ]
  want_seq = seq_df[subjectHits(hits), ]
  temp1 = sequenza_all_samples[match(x, sequenza_all_samples$sampleName), "cellularity"]
  want_mut$sample_id = x
  want_seq$tumor_content = temp1
  final_df = cbind(want_mut[, c("id", "sample_id", "Ref_reads_in_Tumor", "Alt_reads_in_Tumor" )], 
                   want_seq[, c("normal_cn",  "A", "B")])
  colnames(finaldf) = c("mutation_id" , "sample_id", "ref_counts", "var_counts", "normal_cn", 
                        "major_cn","minor_cn")
  write.table(final_df, file.path(resdir_pyclone, paste0(x, "_merged_data.txt")), 
              sep ="\t", quote=FALSE, row.names=FALSE, col.names=T)
})
```


# depth of coverage

Depth of coverage and samtools flagstat was run on each tumor sample and its matched normal
sample in every dataset. The output was combined using the following code for each dataset.
Since the code is the same for each dataset, code below is shown for the dataset with maximum
number of samples ie Quigley et al.

```{r}
rm(list=ls())
dir1="MAIN_DATA/UTR_Project_hg38/david_101/"
setwd(dir1)

flag = list.files(pattern = "flagstat.out", path = "flagstat_res", full.names=T)
doc = list.files(pattern ="*.txt.sample_summary$", path="depth_of_coverage_3utr", full.names = TRUE)
doc2 = list.files( pattern ="*.txt.sample_summary$", path="depth_of_coverage_cds", full.names = TRUE)
doc3 = list.files(pattern = "*.txt.sample_summary$" , path = "depth_of_coverage_5utr", full.names = TRUE)
fl_nms = gsub(".flagstat.out", "", basename(flag))

# depth of coverage FOR 3' UTR
ans1 = sapply(doc, function(x){
  y1=read.delim(x, header=TRUE, stringsAsFactors=FALSE, sep =",")
  y1[1,3]
})
names(ans1) = doc_nms

# depth of coverage FOR 5' UTR
ans_utr5 = sapply(doc3, function(x){
  y1=read.delim(x, header=TRUE, stringsAsFactors=FALSE, sep =",")
  y1[1,3]
})
names(ans1) = doc_nms

# depth of coverage FOR cds
ans2 = sapply(doc2, function(x){
  y1=read.delim(x, header=TRUE, stringsAsFactors=FALSE, sep =",")
  y1[1,3]
})
names(ans2) = doc_nms

# flagstat res.
flag_res=lapply(flag, function(x){
  tryCatch({
    #message(x)
    x1 = read.delim(x, header=FALSE, stringsAsFactors=FALSE)
    x1 = x1[c(1, 4, 5, 9), 1]
    total = as.numeric(unlist(strsplit(x1[1], "[+]"))[1])
    duplicates = as.numeric(unlist(strsplit(x1[2], "[+]"))[1])
    mapped =as.numeric(unlist(strsplit(x1[3], "[+]"))[1])
    properly_paired=as.numeric(unlist(strsplit(x1[4], "[+]"))[1])

    mapped_percent = 100 - (((total-mapped)/total)*100)
    pp_percent = 100- (((total-properly_paired)/total)*100)
    dup_percent =  100- (((total-duplicates)/total)*100)

    c(total, mapped, round(mapped_percent,2),
      properly_paired, round(pp_percent,2),
      duplicates, round(dup_percent,2)  )

  }, warning = function(w) {
    message("warning: ", x)
    c(NA_character_, 7)
  }, error = function(e) {
    c(NA_character_, 7)
  })
})
flag_res =do.call(rbind, flag_res)

colnames(flag_res) = c("total", "mapped", "mapped_percent",
                       "properly_paired", "pp_percent",
                       "duplicates", "dup_percent")
sub_group = "Quigley_et_al"

finaldf = cbind(sampleName= basename(fl_nms),
                depth_of_coverage_3utr = ans1,
                depth_of_coverage_cds = ans2,
                depth_of_coverage_5utr = ans_utr5,
                dataset = rep(sub_group, nrow(flag_res)),
                flag_res)

write.table(finaldf, file.path( "Read_depth_Quigley_results_10_25_2022.txt"),
            sep = "\t", quote=FALSE, row.names=FALSE, col.names=TRUE)
```
# Tumor mutational Burden 
In the code chunk below, we find the TMB for each tumor sample included in our analysis.

```{r}
utr3_width = 46548166
utr5_width = 10209133
cds_width = 35803985


# for 3'utr

df = read.csv("Master_3UTR_Mutations_Hyperremoved_2+.csv")
sampleNames = unique(unlist(strsplit(df$sampleName, ", ")))

mut= lapply(sampleNames, function(x){
  temp = df[grep(paste0("^", x, "$|",  "^", x, ", | ,", x,"$"), df$sampleName), ]
  mut_mb = round(nrow(temp)/  46548166*1000000 , 3)
  c( nrow(temp), mut_mb)
})

ans = cbind(sampleNames, do.call(rbind, mut))
colnames(ans)[2:3] = c("total_utr3_mutations", "mut_mb_utr3")

write.table(ans, "UTR3_TMB_all_patients.txt",
            sep ="\t", quote=FALSE, row.names=FALSE, col.names=TRUE)
# for 5' utr
df = read.csv("Master_5UTR_Mutations_Hyperremoved_2+.csv")
sampleNames = unique(unlist(strsplit(df$sampleName, ", ")))

mut= lapply(sampleNames, function(x){
  temp = df[grep(paste0("^", x, "$|",  "^", x, ", | ,", x,"$"), df$sampleName), ]
  mut_mb = round(nrow(temp)/utr5_width  *1000000 , 3)
  c( nrow(temp), mut_mb)
})
ans = cbind(sampleNames, do.call(rbind, mut))
colnames(ans)[2:3] = c("total_utr5_mutations", "mut_mb_utr5")
write.table(ans, "UTR5_TMB_all_patients.txt",
            sep ="\t", quote=FALSE, row.names=FALSE, col.names=TRUE)

# for CDS
df = read.csv("Master_cds_Mutations_Hyperremoved_2+.csv")
sampleNames = unique(unlist(strsplit(df$sampleName, ", ")))
mut= lapply(sampleNames, function(x){
  temp = df[grep(paste0("^", x, "$|",  "^", x, ", | ,", x,"$"), df$sampleName), ]
  mut_mb = round(nrow(temp)/cds_width  *1000000 , 3)
  c( nrow(temp), mut_mb)
})
ans = cbind(sampleNames, do.call(rbind, mut))
colnames(ans)[2:3] = c("total_cds_mutations", "mut_mb_cds")
write.table(ans, "CDS_TMB_all_patients.txt",
            sep ="\t", quote=FALSE, row.names=FALSE, col.names=TRUE)
```


# base substitution 

```{r}
df = read.csv("Master_3UTR_Mutations_Hyperremoved_2+.csv")

utr5_df = read.delim("Master_5UTR_Mutations_Hyperremoved_2+.txt",
                     header=T, stringsAsFactors = FALSE)
rm = c( grep("-", utr5_df$Alt) , grep("-", utr5_df$Ref) ) # remove indels
length(rm)
utr5_df = utr5_df[ -rm, ]

cds_df = read.delim("Master_cds_Mutations_Hyperremoved_2+.txt", 
                    header=T, stringsAsFactors = FALSE)
rm = c( grep("-", cds_df$Alt) , grep("-", cds_df$Ref) ) # remove indels
rm2 = which(nchar(cds_df$Alt)==2)
length(rm)
cds_df = cds_df[ -c(rm, rm2), ]

met_3utr_base_change = apply(df, 1, function(y) paste0(c(y["Ref"], y["Alt"]), collapse="_"))
met_5utr_base_change = apply(utr5_df, 1, function(y) paste0(c(y["Ref"], y["Alt"]), collapse="_"))
met_cds_base_change = apply(cds_df, 1, function(y) paste0(c(y["Ref"], y["Alt"]), collapse="_"))

pdf("Base_change.pdf", width = 10)
barplot(table(met_3utr_base_change), main = "Met samples :3' utr mutations")
barplot(table(met_5utr_base_change), main = "Met samples :5' utr mutations")
barplot(table(met_cds_base_change), main = "Met samples :cds mutations")
dev.off()

df  = cbind( table(met_3utr_base_change), table(met_5utr_base_change), table(met_cds_base_change))
colnames(df) = c("utr3", "utr5" , "cds")
write.table(df, "data_for_base_change.txt", sep ="\t",
            quote=FALSE ,row.names=TRUE, col.names=TRUE)
```

# cosmic mutational signature analysis

```{r}
rm(list=ls())
library(BSgenome.Hsapiens.UCSC.hg38)
library(deconstructSigs)
library(pheatmap)
library(RColorBrewer)
color =  colorRampPalette(c("grey70", "purple", "black"))(n = 10)

utr3_uncombined =read.csv( "MAIN_DATA/UTR_Project_hg38/Samantha_master_analysis/3UTR_Mutations_Hyperremoved_2+_Uncombined.csv")
cds = read.delim("Master_cds_Mutations_Hyperremoved_2+.txt", header=T, stringsAsFactors = FALSE)

# need to unlist and make uncombined for  cds. 
idx= grep(",", cds$sampleName)
cds_good = cds[-idx, ]
cds_fix = cds[idx, ]
lst = lapply(1:nrow(cds_fix),function(idx){
  x = cds_fix[idx, ]
  samples = unlist(strsplit(unlist(x["sampleName"]), ","))
  y = x[ rep(seq_len(nrow(x)), each = length(samples)), ]
  y$sampleName = samples
  y
})
cds_fix = do.call(rbind, lst)
cds_uncombined = rbind(cds_fix, cds_good)
cds_uncombined$patient_no = 1


utr3_uncombined = utr3_uncombined[, c("Chr", "Start", "End" , "Ref","Alt", "sampleName")]
cds_uncombined = cds_uncombined[, c("Chr", "Start", "End" , "Ref","Alt", "sampleName")]
cds_uncombined$sampleName = trimws(cds_uncombined$sampleName )

# for CDS 

CDS_ans = mut.to.sigs.input(cds_uncombined, sample.id = "sampleName", chr = "Chr",
                        pos = "Start", ref = "Ref", alt = "Alt", bsg = BSgenome.Hsapiens.UCSC.hg38)
UTR3_ans = mut.to.sigs.input(utr3_uncombined, sample.id = "sampleName", chr = "Chr",
                        pos = "Start", ref = "Ref", alt = "Alt", bsg = BSgenome.Hsapiens.UCSC.hg38)
                        
pdf("cds_Mutational_signature_cosmic30.pdf", width = 15)
a1 =lapply(rownames(CDS_ans), function(sample){
  sample_1 = whichSignatures(tumor.ref = ans,
                             signatures.ref = signatures.cosmic,
                             sample.id = sample,
                             contexts.needed = TRUE,
                             tri.counts.method = 'default')


  plotSignatures(sample_1, sub = 'sample')
  sample_1
})
dev.off()

wmat1 = sapply(a1, function(x) x$weights)
wmat1 = as.data.frame(wmat1)
colnames(wmat1) = rownames(ans)
wmat1 = data.matrix(wmat1)

sampleName = colnames(wmat1)
dataset =rep("UW", length(sampleName))
dataset[grep("lucap",  sampleName) ] = "Lucaps"
dataset[grep("DTB",  sampleName) ] = "Quigley"

coldata = data.frame(dataset = dataset, stringsAsFactors = FALSE)
oidx = order(dataset)
ans = ans[oidx, ]
coldata = coldata[ oidx, , drop=FALSE ]
wmat1 = wmat1[, oidx ]
rownames(coldata) = colnames(wmat1)

pdf("cds_cosmic_sig_heatmap.pdf", width =15)
pheatmap(wmat1, scale ="none", cluster_rows=FALSE, cluster_cols = FALSE,
         annotation_col = coldata[, "dataset", drop=FALSE], color = color,
         show_colnames = FALSE)
dev.off()
                       
 
# for 3'utr 

pdf("UTR3_Mutational_signature_cosmic30.pdf", width = 15)
a1 =lapply(rownames(UTR3_ans), function(sample){
  sample_1 = whichSignatures(tumor.ref = ans,
                             signatures.ref = signatures.cosmic,
                             sample.id = sample,
                             contexts.needed = TRUE,
                             tri.counts.method = 'default')


  plotSignatures(sample_1, sub = 'sample')
  sample_1
})
dev.off()

wmat1 = sapply(a1, function(x) x$weights)
wmat1 = as.data.frame(wmat1)
colnames(wmat1) = rownames(ans)
wmat1 = data.matrix(wmat1)

sampleName = colnames(wmat1)
dataset =rep("UW", length(sampleName))
dataset[grep("lucap",  sampleName) ] = "Lucaps"
dataset[grep("DTB",  sampleName) ] = "Quigley"

coldata = data.frame(dataset = dataset, stringsAsFactors = FALSE)
oidx = order(dataset)
ans = ans[oidx, ]
coldata = coldata[ oidx, , drop=FALSE ]
wmat1 = wmat1[, oidx ]
rownames(coldata) = colnames(wmat1)

pdf("UTR3_cosmic_sig_heatmap.pdf", width =15)
pheatmap(wmat1, scale ="none", cluster_rows=FALSE, cluster_cols = FALSE,
         annotation_col = coldata[, "dataset", drop=FALSE], color = color,
         show_colnames = FALSE)
dev.off()
```
                       
