---
title: "SNV analysis"
author: "Sonali Arora"
date: "Nov 16, 2022"
output:
  html_document:
    toc: true
    them

---

# fishHook analysis 

```{r}
rm(list=ls())
library(fishHook)    
library(rtracklayer)
library(org.Hs.eg.db)

expand_df = read.csv("3UTR_Mutations_Hyperremoved_2+_Uncombined.csv", header=T, stringsAsFactors=FALSE)
mutations <- makeGRangesFromDataFrame(expand_df, keep.extra.columns = TRUE)

v22 = import("~/hsiehlab/bioinformatic_references/GATK_hg38_reference/gencode.v22.annotation.gtf")
genes = v22[which(v22$type=="gene"), ]

# calculate the G and C nucleotide frequency.
library(BSgenome.Hsapiens.UCSC.hg38)
genome <- BSgenome.Hsapiens.UCSC.hg38
keep = head(seqlengths(genome), 25) 
gr0 <- GRanges(Rle(c(names(keep))),  
               IRanges(start =1, end =keep))
win = unlist(tile(gr0, width=10001)) 
seq <- getSeq(BSgenome.Hsapiens.UCSC.hg38, win) 
alf <- alphabetFrequency(seq, as.prob=TRUE)[, c("G", "C")] 
head(alf)
colSums(alf)
mcols(win) = alf # add to GRanges object

chromhmm = import('reviewer_questions_March2023/try_hotspot_analysis/BSS01459_18_CALLS_segments.bed.gz') 
hetchromdata = chromhmm %Q% (name %in% c('ZNF/Rpts', 'Het', 'Quies')) 

gc =  fishHook:::Cov(data = win, field = c('C', 'G')) 
hetchrom = Cov(hetchromdata, name = 'Heterochromatin') 

fish3 = fishHook:::Fish(hypotheses = genes,
                        events = mutations,
                        covariates = c( gc, hetchrom),
                        eligible= interval,
                        idcol = 'sampleName',  
                        na.rm = T)

fish3$score()
adv_result <- fish3$res %Q% order(p)
sig_genes2 <- adv_result[which(adv_result$fdr < 0.05),]$gene_name
adv_df <- as.data.frame(adv_result)
write.table(adv_df, file.path("fishHook","fishhook_result_with_gc_content_hetchromm_model_only_utr3_region.txt"), 
            sep="\t", quote=FALSE, row.names=FALSE, col.names=TRUE)
pdf(file.path(resdir, "plot_fishhook_with_gc_content__hetchromm_utr3_region.pdf"), width=10)
fish3$qqp(plotly =FALSE)
dev.off()
```

# oncodriveFML analysis

First we'll liftover coordinates to hg19 
```{r}
df = read.csv("3UTR_Mutations_Hyperremoved_2+_Uncombined.csv")

# liftover coordinates, since ref files are made in hg19
library(rtracklayer)
ch = import("~/hsiehlab/bioinformatic_references/hg38ToHg19.over.chain")
hg38_gr = makeGRangesFromDataFrame(df, keep.extra.columns=T)
hg19_gr = liftOver(hg38_gr, ch)
hg19_gr = unlist(hg19_gr)

hg19_df = as.data.frame(hg19_gr)
write.table(hg19_df, "3UTR_Mutations_Hyperremoved_2+_Uncombined_hg19_liftover.txt", sep ="\t", quote=FALSE, row.names=FALSE, col.names=T)
```
Preparing input files in R 
```{r}
df = read.delim("/fh/fast/hsieh_a/hsiehlab/MAIN_DATA/UTR_Project_hg38/Samantha_master_analysis/3UTR_Mutations_Hyperremoved_2+_Uncombined_hg19_liftover.txt")
df  = df[, c("seqnames", "start", "Ref" , "Alt", "sampleName")]
colnames(df) = c("CHROMOSOME", "POSITION", "REF", "ALT", "SAMPLE")
df[,1] = gsub("chr", "", df[,1])
write.table(df, "MAIN_DATA/oncodrivefml/oncodrivefml_input_all_datasets_hg19.txt", 
            sep="\t", quote=FALSE, row.names=FALSE, col.names=TRUE)


seq_data = read.delim("~/hsiehlab/MAIN_DATA/UTR_Project_hg38/Samantha_master_analysis/sequences/sequences_all_14k_with_location_of_3utr_8_18_2022.txt", 
                      header=T, stringsAsFactors = FALSE)
want = unique(seq_data$utr3_position)
want_id = seq_data[match(want, seq_data$utr3_position), ]$Gene.refGene
seq_gr = as(want, "GRanges")
mcols(seq_gr)$gene_name = want_id 
gr_site_coords = seq_gr
seq_df = as.data.frame(seq_gr)
seq_df = seq_df[, -c(match(c("width", "strand"), colnames(seq_df)))]
colnames(seq_df) = c("CHROMOSOME", "START", "END", "ELEMENT")
write.table(seq_df, "MAIN_DATA/oncodrivefml/elements_utr3_bed_hg19_V2.bed", sep ="\t", quote=F, row.names=FALSE, col.names=T)
```
Running the program :
oncodrivefml -i oncodrivefml_input_all_datasets_hg19.txt -e elements_utr3_bed_hg19_V2.bed

# depth of coverage

Depth of coverage and samtools flagstat was run on each tumor sample and its matched normal
sample in every dataset. The output was combined using the following code for each dataset.
Since the code is the same for each dataset, code below is shown for the dataset with maximum
number of samples ie Quigley et al.

```{r}
rm(list=ls())
dir1="MAIN_DATA/UTR_Project_hg38/david_101/"
setwd(dir1)

flag = list.files(pattern = "flagstat.out", path = "flagstat_res", full.names=T)
doc = list.files(pattern ="*.txt.sample_summary$", path="depth_of_coverage_3utr", full.names = TRUE)
doc2 = list.files( pattern ="*.txt.sample_summary$", path="depth_of_coverage_cds", full.names = TRUE)
doc3 = list.files(pattern = "*.txt.sample_summary$" , path = "depth_of_coverage_5utr", full.names = TRUE)
fl_nms = gsub(".flagstat.out", "", basename(flag))

# depth of coverage FOR 3' UTR
ans1 = sapply(doc, function(x){
  y1=read.delim(x, header=TRUE, stringsAsFactors=FALSE, sep =",")
  y1[1,3]
})
names(ans1) = doc_nms

# depth of coverage FOR 5' UTR
ans_utr5 = sapply(doc3, function(x){
  y1=read.delim(x, header=TRUE, stringsAsFactors=FALSE, sep =",")
  y1[1,3]
})
names(ans1) = doc_nms

# depth of coverage FOR cds
ans2 = sapply(doc2, function(x){
  y1=read.delim(x, header=TRUE, stringsAsFactors=FALSE, sep =",")
  y1[1,3]
})
names(ans2) = doc_nms

# flagstat res.
flag_res=lapply(flag, function(x){
  tryCatch({
    #message(x)
    x1 = read.delim(x, header=FALSE, stringsAsFactors=FALSE)
    x1 = x1[c(1, 4, 5, 9), 1]
    total = as.numeric(unlist(strsplit(x1[1], "[+]"))[1])
    duplicates = as.numeric(unlist(strsplit(x1[2], "[+]"))[1])
    mapped =as.numeric(unlist(strsplit(x1[3], "[+]"))[1])
    properly_paired=as.numeric(unlist(strsplit(x1[4], "[+]"))[1])

    mapped_percent = 100 - (((total-mapped)/total)*100)
    pp_percent = 100- (((total-properly_paired)/total)*100)
    dup_percent =  100- (((total-duplicates)/total)*100)

    c(total, mapped, round(mapped_percent,2),
      properly_paired, round(pp_percent,2),
      duplicates, round(dup_percent,2)  )

  }, warning = function(w) {
    message("warning: ", x)
    c(NA_character_, 7)
  }, error = function(e) {
    c(NA_character_, 7)
  })
})
flag_res =do.call(rbind, flag_res)

colnames(flag_res) = c("total", "mapped", "mapped_percent",
                       "properly_paired", "pp_percent",
                       "duplicates", "dup_percent")
sub_group = "Quigley_et_al"

finaldf = cbind(sampleName= basename(fl_nms),
                depth_of_coverage_3utr = ans1,
                depth_of_coverage_cds = ans2,
                depth_of_coverage_5utr = ans_utr5,
                dataset = rep(sub_group, nrow(flag_res)),
                flag_res)

write.table(finaldf, file.path( "Read_depth_Quigley_results_10_25_2022.txt"),
            sep = "\t", quote=FALSE, row.names=FALSE, col.names=TRUE)
```
# Tumor mutational Burden 
In the code chunk below, we find the TMB for each tumor sample included in our analysis.

```{r}
utr3_width = 46548166
utr5_width = 10209133
cds_width = 35803985


# for 3'utr

df = read.csv("Master_3UTR_Mutations_Hyperremoved_2+.csv")
sampleNames = unique(unlist(strsplit(df$sampleName, ", ")))

mut= lapply(sampleNames, function(x){
  temp = df[grep(paste0("^", x, "$|",  "^", x, ", | ,", x,"$"), df$sampleName), ]
  mut_mb = round(nrow(temp)/  46548166*1000000 , 3)
  c( nrow(temp), mut_mb)
})

ans = cbind(sampleNames, do.call(rbind, mut))
colnames(ans)[2:3] = c("total_utr3_mutations", "mut_mb_utr3")

write.table(ans, "UTR3_TMB_all_patients.txt",
            sep ="\t", quote=FALSE, row.names=FALSE, col.names=TRUE)
# for 5' utr
df = read.csv("Master_5UTR_Mutations_Hyperremoved_2+.csv")
sampleNames = unique(unlist(strsplit(df$sampleName, ", ")))

mut= lapply(sampleNames, function(x){
  temp = df[grep(paste0("^", x, "$|",  "^", x, ", | ,", x,"$"), df$sampleName), ]
  mut_mb = round(nrow(temp)/utr5_width  *1000000 , 3)
  c( nrow(temp), mut_mb)
})
ans = cbind(sampleNames, do.call(rbind, mut))
colnames(ans)[2:3] = c("total_utr5_mutations", "mut_mb_utr5")
write.table(ans, "UTR5_TMB_all_patients.txt",
            sep ="\t", quote=FALSE, row.names=FALSE, col.names=TRUE)

# for CDS
df = read.csv("Master_cds_Mutations_Hyperremoved_2+.csv")
sampleNames = unique(unlist(strsplit(df$sampleName, ", ")))
mut= lapply(sampleNames, function(x){
  temp = df[grep(paste0("^", x, "$|",  "^", x, ", | ,", x,"$"), df$sampleName), ]
  mut_mb = round(nrow(temp)/cds_width  *1000000 , 3)
  c( nrow(temp), mut_mb)
})
ans = cbind(sampleNames, do.call(rbind, mut))
colnames(ans)[2:3] = c("total_cds_mutations", "mut_mb_cds")
write.table(ans, "CDS_TMB_all_patients.txt",
            sep ="\t", quote=FALSE, row.names=FALSE, col.names=TRUE)
```


# base substitution 

```{r}
df = read.csv("Master_3UTR_Mutations_Hyperremoved_2+.csv")

utr5_df = read.delim("Master_5UTR_Mutations_Hyperremoved_2+.txt",
                     header=T, stringsAsFactors = FALSE)
rm = c( grep("-", utr5_df$Alt) , grep("-", utr5_df$Ref) ) # remove indels
length(rm)
utr5_df = utr5_df[ -rm, ]

cds_df = read.delim("Master_cds_Mutations_Hyperremoved_2+.txt", 
                    header=T, stringsAsFactors = FALSE)
rm = c( grep("-", cds_df$Alt) , grep("-", cds_df$Ref) ) # remove indels
rm2 = which(nchar(cds_df$Alt)==2)
length(rm)
cds_df = cds_df[ -c(rm, rm2), ]

met_3utr_base_change = apply(df, 1, function(y) paste0(c(y["Ref"], y["Alt"]), collapse="_"))
met_5utr_base_change = apply(utr5_df, 1, function(y) paste0(c(y["Ref"], y["Alt"]), collapse="_"))
met_cds_base_change = apply(cds_df, 1, function(y) paste0(c(y["Ref"], y["Alt"]), collapse="_"))

pdf("Base_change.pdf", width = 10)
barplot(table(met_3utr_base_change), main = "Met samples :3' utr mutations")
barplot(table(met_5utr_base_change), main = "Met samples :5' utr mutations")
barplot(table(met_cds_base_change), main = "Met samples :cds mutations")
dev.off()

df  = cbind( table(met_3utr_base_change), table(met_5utr_base_change), table(met_cds_base_change))
colnames(df) = c("utr3", "utr5" , "cds")
write.table(df, "data_for_base_change.txt", sep ="\t",
            quote=FALSE ,row.names=TRUE, col.names=TRUE)
```

# cosmic mutational signature analysis

```{r}
rm(list=ls())
library(BSgenome.Hsapiens.UCSC.hg38)
library(deconstructSigs)
library(pheatmap)
library(RColorBrewer)
color =  colorRampPalette(c("grey70", "purple", "black"))(n = 10)

utr3_uncombined =read.csv( "MAIN_DATA/UTR_Project_hg38/Samantha_master_analysis/3UTR_Mutations_Hyperremoved_2+_Uncombined.csv")
cds = read.delim("Master_cds_Mutations_Hyperremoved_2+.txt", header=T, stringsAsFactors = FALSE)

# need to unlist and make uncombined for  cds. 
idx= grep(",", cds$sampleName)
cds_good = cds[-idx, ]
cds_fix = cds[idx, ]
lst = lapply(1:nrow(cds_fix),function(idx){
  x = cds_fix[idx, ]
  samples = unlist(strsplit(unlist(x["sampleName"]), ","))
  y = x[ rep(seq_len(nrow(x)), each = length(samples)), ]
  y$sampleName = samples
  y
})
cds_fix = do.call(rbind, lst)
cds_uncombined = rbind(cds_fix, cds_good)
cds_uncombined$patient_no = 1


utr3_uncombined = utr3_uncombined[, c("Chr", "Start", "End" , "Ref","Alt", "sampleName")]
cds_uncombined = cds_uncombined[, c("Chr", "Start", "End" , "Ref","Alt", "sampleName")]
cds_uncombined$sampleName = trimws(cds_uncombined$sampleName )

# for CDS 

CDS_ans = mut.to.sigs.input(cds_uncombined, sample.id = "sampleName", chr = "Chr",
                        pos = "Start", ref = "Ref", alt = "Alt", bsg = BSgenome.Hsapiens.UCSC.hg38)
UTR3_ans = mut.to.sigs.input(utr3_uncombined, sample.id = "sampleName", chr = "Chr",
                        pos = "Start", ref = "Ref", alt = "Alt", bsg = BSgenome.Hsapiens.UCSC.hg38)
                        
pdf("cds_Mutational_signature_cosmic30.pdf", width = 15)
a1 =lapply(rownames(CDS_ans), function(sample){
  sample_1 = whichSignatures(tumor.ref = ans,
                             signatures.ref = signatures.cosmic,
                             sample.id = sample,
                             contexts.needed = TRUE,
                             tri.counts.method = 'default')


  plotSignatures(sample_1, sub = 'sample')
  sample_1
})
dev.off()

wmat1 = sapply(a1, function(x) x$weights)
wmat1 = as.data.frame(wmat1)
colnames(wmat1) = rownames(ans)
wmat1 = data.matrix(wmat1)

sampleName = colnames(wmat1)
dataset =rep("UW", length(sampleName))
dataset[grep("lucap",  sampleName) ] = "Lucaps"
dataset[grep("DTB",  sampleName) ] = "Quigley"

coldata = data.frame(dataset = dataset, stringsAsFactors = FALSE)
oidx = order(dataset)
ans = ans[oidx, ]
coldata = coldata[ oidx, , drop=FALSE ]
wmat1 = wmat1[, oidx ]
rownames(coldata) = colnames(wmat1)

pdf("cds_cosmic_sig_heatmap.pdf", width =15)
pheatmap(wmat1, scale ="none", cluster_rows=FALSE, cluster_cols = FALSE,
         annotation_col = coldata[, "dataset", drop=FALSE], color = color,
         show_colnames = FALSE)
dev.off()
                       
 
# for 3'utr 

pdf("UTR3_Mutational_signature_cosmic30.pdf", width = 15)
a1 =lapply(rownames(UTR3_ans), function(sample){
  sample_1 = whichSignatures(tumor.ref = ans,
                             signatures.ref = signatures.cosmic,
                             sample.id = sample,
                             contexts.needed = TRUE,
                             tri.counts.method = 'default')


  plotSignatures(sample_1, sub = 'sample')
  sample_1
})
dev.off()

wmat1 = sapply(a1, function(x) x$weights)
wmat1 = as.data.frame(wmat1)
colnames(wmat1) = rownames(ans)
wmat1 = data.matrix(wmat1)

sampleName = colnames(wmat1)
dataset =rep("UW", length(sampleName))
dataset[grep("lucap",  sampleName) ] = "Lucaps"
dataset[grep("DTB",  sampleName) ] = "Quigley"

coldata = data.frame(dataset = dataset, stringsAsFactors = FALSE)
oidx = order(dataset)
ans = ans[oidx, ]
coldata = coldata[ oidx, , drop=FALSE ]
wmat1 = wmat1[, oidx ]
rownames(coldata) = colnames(wmat1)

pdf("UTR3_cosmic_sig_heatmap.pdf", width =15)
pheatmap(wmat1, scale ="none", cluster_rows=FALSE, cluster_cols = FALSE,
         annotation_col = coldata[, "dataset", drop=FALSE], color = color,
         show_colnames = FALSE)
dev.off()
```
                       
