---
title: "Clustering_Significance"
output: html_document
Author: Samantha L. Schuster
date: "2022-11-14"
---


```{r}
library(tidyverse)
library(hrbrthemes)
```

```{r import}
utr3_muts <- read_csv("3UTR_Mutations_Hyperremoved_2+_Uncombined.csv") %>%
  select(Chr, Start, Gene.refGene, sampleName, id)

utr3_combined <- read_csv("Master_3UTR_Mutations_Hyperremoved_2+.csv")
```

Find mutations near each other (within 50bp)

```{r}
utr3_genes <- split(utr3_muts, utr3_muts$Gene.refGene)

pairs <- data.frame(matrix(ncol = 4, nrow = 0))
x <- c("Gene", "Loc1", "Loc2", "Dist")
colnames(pairs) <- x

for(gene in utr3_genes){
  chr <- unique(gene$Chr)
  if (length(unique(gene$Start))>1){
    locs <- unique(gene$Start)
    for(i in (1:(length(locs)-1))){
      for(j in (i+1):length(locs)){
        if(abs(locs[j]-locs[i])<50){
          pair <- c(unique(gene$Gene.refGene),paste0(chr, ":",locs[i]),paste0(chr, ":",locs[j]), abs(locs[j]-locs[i]))
          pairs[nrow(pairs)+ 1, ] <- pair}
    }
  }
  
  }}

pairs$Dist <- as.numeric(pairs$Dist)
```

Distribution of clustered distances

```{r}
pairs <- read_csv("Hotspot mutations.csv")

ggplot(pairs, aes(x=Dist)) + 
  geom_histogram(binwidth= 5, fill="#86608e", color="#e9ecef", alpha=0.9) + 
  theme_minimal() + 
  labs(x="Distance between Mutations", y="Count", title = "Hotspot mutation distance")
```

Compare to MPRA sig mutations

```{r}
#MPRA results
IVT_MPRA <- read_csv("Read_count_filtered/high_filters_downstream.csv")[,-1]
poly_MPRA <- read_csv("All_Polysome_MPRA_results.csv")[,-c(1,3:5,10:13)]

IVT_sig <- IVT_MPRA %>% filter(sig_E == "Sig") %>% pull(barcode)
MPRA_sig <- unique(poly_MPRA %>% filter(xtail_FDR_v2<0.1, Comparison %in% c("HighPoly:RNA","TotalPoly:RNA")) %>% pull(barcode))

IVT_sig <- substring(IVT_sig, 1, nchar(IVT_sig)-4)
MPRA_sig <- substring(MPRA_sig, 1, nchar(MPRA_sig)-4)

recurrent <- utr3_combined %>% filter(patient_no>1) %>% pull(id)
recurrent <- substring(recurrent, 1, nchar(recurrent)-4)

hotspots <- unique(c(pairs$Loc1, pairs$Loc2, recurrent))

intersect(IVT_sig, hotspots)
intersect(MPRA_sig, hotspots)
```
FishHook analysis: 

```{r}
rm(list=ls())
library(fishHook)    
library(rtracklayer)
library(org.Hs.eg.db)

expand_df = read.csv("3UTR_Mutations_Hyperremoved_2+_Uncombined.csv", header=T, stringsAsFactors=FALSE)
mutations <- makeGRangesFromDataFrame(expand_df, keep.extra.columns = TRUE)

v22 = import("~/hsiehlab/bioinformatic_references/GATK_hg38_reference/gencode.v22.annotation.gtf")
genes = v22[which(v22$type=="gene"), ]

# calculate the G and C nucleotide frequency.
library(BSgenome.Hsapiens.UCSC.hg38)
genome <- BSgenome.Hsapiens.UCSC.hg38
keep = head(seqlengths(genome), 25) 
gr0 <- GRanges(Rle(c(names(keep))),  
               IRanges(start =1, end =keep))
win = unlist(tile(gr0, width=10001)) 
seq <- getSeq(BSgenome.Hsapiens.UCSC.hg38, win) 
alf <- alphabetFrequency(seq, as.prob=TRUE)[, c("G", "C")] 
head(alf)
colSums(alf)
mcols(win) = alf # add to GRanges object

chromhmm = import('reviewer_questions_March2023/try_hotspot_analysis/BSS01459_18_CALLS_segments.bed.gz') 
hetchromdata = chromhmm %Q% (name %in% c('ZNF/Rpts', 'Het', 'Quies')) 

gc =  fishHook:::Cov(data = win, field = c('C', 'G')) 
hetchrom = Cov(hetchromdata, name = 'Heterochromatin') 

fish3 = fishHook:::Fish(hypotheses = genes,
                        events = mutations,
                        covariates = c( gc, hetchrom),
                        eligible= interval,
                        idcol = 'sampleName',  
                        na.rm = T)

fish3$score()
adv_result <- fish3$res %Q% order(p)
sig_genes2 <- adv_result[which(adv_result$fdr < 0.05),]$gene_name
adv_df <- as.data.frame(adv_result)
write.table(adv_df, file.path("fishHook","fishhook_result_with_gc_content_hetchromm_model_only_utr3_region.txt"), 
            sep="\t", quote=FALSE, row.names=FALSE, col.names=TRUE)
pdf(file.path(resdir, "plot_fishhook_with_gc_content__hetchromm_utr3_region.pdf"), width=10)
fish3$qqp(plotly =FALSE)
dev.off()
```

Make lollipop plots

```{r}
sig_hot <- read_csv("Hotspots/Significant_hotspots.csv")

temp <- sig_hot %>% filter(Gene == "CCL11")
tstart <- temp$start[1]
tend <- temp$end[1]

CCL11_lolli <- ggplot(temp, aes(x=position, y=patient_no)) +
  geom_segment(aes(x=position, xend=position, y=0, yend=patient_no), color = "grey") +
  geom_point(color = "slateblue", size = 4) + 
  theme_minimal() +
  coord_cartesian(xlim=c(tstart, tend)) + 
  labs(x="Chromosomal position", y="# patients", title = "CCL11 Hotspot")

```

